src/Type/SchemaValidationContext.php

```diff
@@ -694,23 +716,25 @@ private function getDirectives($object)
     }
 
     /**
-     * @param InterfaceType $iface
+     * @param ObjectType|InterfaceType $type
+     * @param InterfaceType            $iface
      *
      * @return NamedTypeNode|null
      */
-    private function getImplementsInterfaceNode(ObjectType $type, $iface)
+    private function getImplementsInterfaceNode($type, InterfaceType $iface) : ?NamedTypeNode
```

Comment:
@spawnia This function is used in case of interface extension errors in SchemaValdationContext.php:676 (`validateInterfaces`) which causes a type error in our tests. There is a test that specifically checks that a schema where an object implements a non-interface doesn't validate (that scenario would hit this code path).

I think this is a bug in the `validateObjectInterfaces` implementation which shouldn't call `getImplementsInterfaceNodes` there with a known non-interface. However, I'm unsure what the context for `reportError` in terms of `$nodes` should be instead.

Reply:
The only thing `getAllImplementsInterfaceNodes` actually needs from the `$iface`-like object is a name property. The method was providing the relevant `NamedTypeNode` instance (rather than not returning anything at all) so removing that context would be a step backwards.

I've widened the type to `Type` to make a step in the direction of type safety without breaking existing functionality. I've had some other places where I needed `ImplementingType` but also needed a `name` property and thus had to fallback to an annotation of `InterfaceType | ObjectType`. A good follow-up to this issue would be to introduce a generic `TypeInterface` which defines things shared by all type implementations. The `ImplementingInterface` could extend this so we can accept that interface directly instead of a union of interface and object. (That's not needed in PHP 8 where we could just do `ObjectType | InterfaceType` which may be more desirable since it's debatable whether custom implementations of those types are useful 🤷‍♂️ )

Reply:
Alright. We can stick with `Type`, that works fine. I adapted the naming to indicate it is not always an `InterfaceType`.